pkg/kubeapiserver/authenticator/config.go

```diff
@@ -224,7 +241,49 @@ func (config Config) New() (authenticator.Request, *spec.SecurityDefinitions, sp
 		authenticator = union.NewFailOnError(authenticator, anonymous.NewAuthenticator())
 	}
 
-	return authenticator, &securityDefinitionsV2, securitySchemesV3, nil
+	return authenticator, updateAuthenticationConfig, &securityDefinitionsV2, securitySchemesV3, nil
+}
+
+type jwtAuthenticatorWithCancel struct {
+	jwtAuthenticator authenticator.Token
+	cancel           func()
+}
+
+func newJWTAuthenticator(ctx context.Context, config *apiserver.AuthenticationConfiguration, oidcSigningAlgs []string, apiAudiences authenticator.Audiences, disallowedIssuers []string, synchronousInitialization bool) (_ *jwtAuthenticatorWithCancel, buildErr error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	defer func() {
+		if buildErr != nil {
+			cancel()
+		}
+	}()
+	var jwtAuthenticators []authenticator.Token
+	for _, jwtAuthenticator := range config.JWT {
+		// TODO remove this CAContentProvider indirection
+		var oidcCAContent oidc.CAContentProvider
+		if len(jwtAuthenticator.Issuer.CertificateAuthority) > 0 {
+			var oidcCAError error
+			oidcCAContent, oidcCAError = dynamiccertificates.NewStaticCAContent("oidc-authenticator", []byte(jwtAuthenticator.Issuer.CertificateAuthority))
+			if oidcCAError != nil {
+				return nil, oidcCAError
+			}
+		}
+		oidcAuth, err := oidc.New(ctx, oidc.Options{
+			JWTAuthenticator:          jwtAuthenticator,
+			CAContentProvider:         oidcCAContent,
+			SupportedSigningAlgs:      oidcSigningAlgs,
+			DisallowedIssuers:         disallowedIssuers,
+			SynchronousInitialization: synchronousInitialization,
+		})
+		if err != nil {
+			return nil, err
+		}
+		jwtAuthenticators = append(jwtAuthenticators, oidcAuth)
+	}
+	return &jwtAuthenticatorWithCancel{
+		jwtAuthenticator: authenticator.WrapAudienceAgnosticToken(apiAudiences, tokenunion.NewFailOnError(jwtAuthenticators...)), // this handles the empty jwtAuthenticators slice case correctly
```

Comment:
> This looks like a behavior change.

We never allowed more than on jwt authenticator before, so the distinction between these two never actually came up.

> Were we doing the wrong thing before?

Yeah, because we know that each jwt authenticator is seperated from the rest via issuer, and we check that immediately before doing anything else, and when the issuer doesn't match we return `nil, false, nil`.  So once you hit an error case, there is no reason to keep going to another jwt authenticator.

> Seems like this needs a test exercising multiple authenticators and error-continuing behavior whichever way we decide

👍

Reply:
> Seems like this needs a test exercising multiple authenticators and error-continuing behavior whichever way we decide

Actually I think the existing `TestMultipleJWTAuthenticators` covers this already.